esphome/components/tuya/number/__init__.py

```diff
@@ -81,5 +85,5 @@ async def to_code(config):
     cg.add(var.set_number_id(config[CONF_NUMBER_DATAPOINT]))
     if hidden_config := config.get(CONF_DATAPOINT_HIDDEN):
         cg.add(var.set_datapoint_type(hidden_config[CONF_DATAPOINT_TYPE]))
-        if hidden_init_config := hidden_config.get(CONF_INIT):
-            cg.add(var.set_datapoint_restore_value(hidden_init_config[CONF_VALUE]))
+        if hidden_init_value := hidden_config.get(CONF_INITIAL_VALUE):
```

Comment:
Gotta protect against 0 being falsy here
```suggestion
        if (hidden_init_value := hidden_config.get(CONF_INITIAL_VALUE, None)) is not None:
```